### ISSUE TITLE
Incorrect unreachable code problem

### ISSUE THREAD

**Issue by odwrotnie** [scalacenter/bloop#1495]
Bloop is reporting `unreachable code` in wrong place

# The code

```
sealed trait ItemsEventApi extends InstapEventApi {
  def timestamp: Instant
}
object ItemsEventApi {
  implicit val format: Format[ItemsEventApi] = Json.format
}
```

# Visual Studio Code screenshot

![image](https://user-images.githubusercontent.com/473035/116139761-a9634700-a6d6-11eb-8d96-dee8f6d1557b.png)

**Comment by tgodzik** [comment]
Thanks for reporting! Do you maybe have a smaller reproduction that we can take a look at? Is it just a question of wrong position, stale warning or altogether out of the blue weird warning? Does it get reported via the build tool itself?

The warning comes from the compiler itself, but some things might be broken when it comes to incremental compilation. Did you try clean build?

**Comment by odwrotnie** [comment]
It is a weird warning in this class only. I have tried to clean.

**Comment by tgodzik** [comment]
Does it pop up with sbt? Any chance for a reproduction?

**Comment by odwrotnie** [comment]
It is only visible in VS Code, SBT does not mention it.

**Comment by jozanek** [comment]
**TL;DR: not a Bloop bug.** The warning is a real scalac diagnostic caused by play-json's
Scala 2 macro (≤ 2.9.x): `JsMacroImpl.directKnownSubclasses` collects the sealed family's
subtypes — intermediate traits included — into an unordered `Set[Type]`, and generates the
`Writes` match in Set iteration order. If a supertype case precedes one of its own subtypes,
the later case is genuinely unreachable, and scalac warns at the `Json.format` call — exactly
the screenshot.

That order is deterministic per compilation context but flips on semantically irrelevant
changes (I could toggle the warning just by adding unused `class PaddingN` declarations), so
sbt and Bloop can legitimately disagree on the same sources. Bloop also republishes
analysis-persisted warnings to BSP clients on a new session (intended behavior), which keeps
it pinned in VS Code, while sbt only prints warnings for files it recompiles. `-Yrangepos` was
ruled out.

Repro (Scala 2.13, play-json 2.9.2, plain scalac — no Bloop):

```scala
import play.api.libs.json._

object Models {
  sealed trait Event
  trait ItemEvent extends Event { def id: String }
  case class ItemAdded(id: String) extends ItemEvent
  case class ItemRemoved(id: String) extends ItemEvent

  object ItemAdded { implicit val f: OFormat[ItemAdded] = Json.format }
  object ItemRemoved { implicit val f: OFormat[ItemRemoved] = Json.format }
  object ItemEvent {
    implicit val f: OFormat[ItemEvent] =
      OFormat(Reads[ItemEvent](_ => JsError("no")), OWrites[ItemEvent](_ => Json.obj()))
  }
  object Event { implicit val f: OFormat[Event] = Json.format } // <- "unreachable code"
}
```

Declare `trait ItemEvent` *after* the case classes and the warning disappears.

Upstream confirmed and fixed the nondeterminism in playframework/play-json#1047 /
playframework/play-json#1052, released in 2.10.6 / 3.0.4.

**Suggestion:** close as not-a-Bloop-bug; affected users should upgrade play-json to
≥ 2.10.6 / 3.0.4, or silence the (possibly legitimate) warning with `-Wconf`.

**Comment by tgodzik** [comment]
Thanks for investigating @jozanek ! Since we don't have a reproduction it's plausible enough that it's the problem you mention.
